### DISCUSSION TITLE
WebAssembly.Function.type not on some platforms

### DISCUSSION THREAD

**Discussion by rentallect** [emscripten-core/emscripten#20904] [Q&A] [answered]
I have some JSPI-enabled WASM constructed with `emscripten` `3.1.46`.  Even when the consuming browser has the JSPI feature enabled, there are some platforms that experience exceptions during the WASM instantiation because the `emscripten loader code` invokes `WebAssembly.Function.type` and the `WebAssembly.Function` object doesn't have `type`.

For example Mac 12.7 has `WebAssembly.Function.type`, but...

Mac 14.x doesn't.  
Newer Windows11 doesn't.  
Newer Fedora doesn't.
...maybe more

Can someone explain what I might be missing here, and if there is a way to resolve this (as in, newer emscripten releases might have fixed this)?

**Comment by kripken** [comment]
@rentallect That sounds like missing browser support. It might be helpful to know which browsers and which versions you see the issue in (that would be more specific than operating system information).

But @brendandahl who added the usage of `WebAssembly.Function.type` in #20401 might know the answer - from that PR it sounds like there were some changes to the type reflection API and perhaps those haven't reached all browsers yet?

**Reply by rentallect** [reply]
I'll get you more data (browser versions affected), but the interesting thing is that `WebAssembly.Function.type` seems to be vanishing from newer/updated platforms.  That is, the same WASM which was working on certain platforms suddenly stopped working after users updated their operating systems.  `WebAssembly.Function` is there, but `type` is now gone.

**Reply by rentallect** [reply]
I updated my Chrome today.  I now have `Version 120.0.6099.109 (Official Build) (arm64)`.  Prior to this update, `WebAssembly.Function.type` was there, but now it's gone in Chromium, and as a result, my WASM will no longer run.

**Reply by sbc100** [reply]
From the comment it looks like `original.type()` the new thing and `WebAssembly.Function.type` is the old thing, but that code expects one or the other to exist.

See https://github.com/emscripten-core/emscripten/blob/67774cf6d27bf12afd1266c3d48fb60dd8d576e2/src/library_async.js#L457-L458

However, it looks like 3.1.46 doesn't have that fallback:  https://github.com/emscripten-core/emscripten/blob/19607820c447a13fd8d0b7680c56148427d6e1b8/src/library_async.js#L461

So it looks like the solution here is rebuild using a newer version of emscripten that supports the newer shape of this API.   
 
This kind of breakage is, I'm afraid, sometimes unavoidable when using new and experimental APIs like JSPI.

(@brendandahl is my assessment correct here?)

**Reply by rentallect** [reply, marked as answer]
I'll upgrade to `3.1.51`, rebuild everything, and then report back with the results.
